### PR TITLE
Smoother AI tracking

### PR DESCRIPTION
**Pull request by SECBATON-GRIFFON**
[tweak]

## What this does
redoes #34059 but without the retracking stuff as it caused some controversy
actually locks the AI eye to the tracked mob so it gets updated on movement rather than every second.
adds an event that detects if someone ~~enters or~~ exits the cameranet on a turf to check if the AI can keep following them through static instead of checking every second. (with the usual exception for borgs with a camera) ~~does not apply when tracking stops via usual means like moving away or jumping somewhere else.~~
changes the snowflaked checks for AI blocking items to be a variable called "blocks_tracking" that does its functionality.

## Why it's good
less jumpy when tracking mobs

## Changelog
:cl:
 * tweak: AIs can now track mobs as they move, rather than jumping to them every second.
 * tweak: AI messages when they stop being able to follow a target are now more general, as cameras are not the only thing that stops tracking.

### CODE REVIEW

**Comment by SonixApache** [comment]
has your parole been lifted yet?

File: code/_hooks/events.dm [hunk 1]
```diff
@@ -262,6 +262,9 @@
 // atom/movable/exiter: the movable exiting the area
 /event/area_exited
 
+// Called by miscellaneous functions not covered by entered, equipped and unequipped events for cameranet updates
+/event/cameranet_changed
```
Comment (jwhitak) [inline code/_hooks/events.dm]: Change this variable name and description - cameranets are defined as camera networks, such as ss13 main station, security, outpost, etc. This event tracks things that would add/remove camera tracking specifically, and should have a relevant name, such as camera_tracking_change or whatnot.

Additionally, try to document the arguments used.

File: code/game/machinery/camera/tracking.dm [hunk 2]
```diff
@@ -183,34 +181,48 @@
...
+
+/mob/living/silicon/ai/proc/on_camera_change(mover)
+	if(eyeobj && currently_tracking)
+		var/cantrack = can_track_atom(currently_tracking)
+		/*if(!eyeobj.locked_to && cantrack) // Retracking code
+			to_chat(src, "Target is trackable again.")
+			currently_tracking.lock_atom(eyeobj,/datum/locking_category/ai_eye)
+		else */if(!cantrack && eyeobj.locked_to == currently_tracking)
+			to_chat(src, "Target is no longer trackable.")
+			currently_tracking.unlock_atom(eyeobj)
```
Comment (jwhitak) [inline code/game/machinery/camera/tracking.dm]: This is called in stop_ai_tracking, why is it called here?

File: code/game/machinery/camera/tracking.dm [hunk 3]
```diff
@@ -159,18 +162,13 @@
...
-		if(target_human.is_wearing_item(/obj/item/clothing/gloves/ninja))
-			return FALSE
-		if(target_human.is_wearing_item(/obj/item/clothing/head/tinfoil))
-			return FALSE
-		if(target_human.is_holding_item(/obj/item/device/megaphone/madscientist))
-			return FALSE
+		for(var/obj/item/I in target_mob.get_all_slots())
+			if(I.blocks_tracking)
+				return FALSE
+			if(target_mob.is_wearing_item(I,slot_wear_id))
```
Comment (jwhitak) [inline code/game/machinery/camera/tracking.dm:168]: Doesn't get_all_slots return the ID slot already?

File: code/game/machinery/camera/tracking.dm [hunk 4]
```diff
@@ -183,34 +181,48 @@
...
+
+	currently_tracking.register_event(/event/moved,src,nameof(src::on_camera_change()))
+	currently_tracking.register_event(/event/equipped,src,nameof(src::on_camera_change()))
+	//currently_tracking.register_event(/event/unequipped,src,nameof(src::on_camera_change()))
+	currently_tracking.register_event(/event/cameranet_changed,src,nameof(src::on_camera_change()))
+	to_chat(src, "Now tracking [currently_tracking.name] on camera.")
+
+/datum/locking_category/ai_eye
+
+/mob/living/silicon/ai/proc/on_camera_change(mover)
```
Comment (jwhitak) [inline code/game/machinery/camera/tracking.dm]: Is the mover variable used at any point?

File: code/modules/mob/living/silicon/ai/freelook/eye.dm [hunk 5]
```diff
@@ -122,6 +121,8 @@
 		CAN_MOVE_DIAGONALLY = FALSE
 		user.eyeobj.glide_size = DELAY2GLIDESIZE(1)
 		user.delayNextMove(1)
+	else if(user.currently_tracking)
+		user.eyeobj.glide_size = user.currently_tracking.glide_size
```
Comment (jwhitak) [inline code/modules/mob/living/silicon/ai/freelook/eye.dm]: Glide size doesn't seem to do anything unless the AI actively has the client call Move, so this doesn't really work when tracking from what I can tell

**Review by jwhitak — CHANGES_REQUESTED** [review]
There are multiple issues with this PR, and in my testing I found at least one more problem.

```
Now tracking Ian on camera.
No longer tracking Ian on camera.
Now tracking Dessert on camera.
ADMIN LOG: [Dilt]/(H.E.R.B.I.E.) deleted Dessert at (264,257,1)
```
I had jumped on the old local test server, used the track button to look at Ian, then used the button to look at Dessert. I qdel'd Dessert, and after a moment, the camera suddenly jumped back to Ian.

And when tracking Ian, it wasn't even smooth. Just an instant jump. I saw the step changes to try and help the tracking, but it doesn't seem to work?

File: code/game/machinery/camera/tracking.dm (hunk 4 above)
Comment (SECBATON-GRIFFON) [reply code/game/machinery/camera/tracking.dm]: nah, just needed so the event can be invoked, changed it to after_move which doesnt

File: code/modules/mob/living/silicon/ai/freelook/eye.dm (hunk 5 above)
Comment (SECBATON-GRIFFON) [reply code/modules/mob/living/silicon/ai/freelook/eye.dm]: i have noticed this in testing, removed this

File: code/game/machinery/camera/tracking.dm (hunk 2 above)
Comment (SECBATON-GRIFFON) [reply code/game/machinery/camera/tracking.dm]: forgot to comment it out, fixed

File: code/_hooks/events.dm (hunk 1 above)
Comment (SECBATON-GRIFFON) [reply code/_hooks/events.dm]: used camera_sight_changed instead and documented them, done

File: code/game/machinery/camera/tracking.dm (hunk 3 above)
Comment (SECBATON-GRIFFON) [reply code/game/machinery/camera/tracking.dm:168]: nah, just the items from them, not their positions on the mob

**Comment by SECBATON-GRIFFON** [comment]
> There are multiple issues with this PR, and in my testing I found at least one more problem.
> 
> ```
> Now tracking Ian on camera.
> No longer tracking Ian on camera.
> Now tracking Dessert on camera.
> ADMIN LOG: [Dilt]/(H.E.R.B.I.E.) deleted Dessert at (264,257,1)
> ```
> 
> I had jumped on the old local test server, used the track button to look at Ian, then used the button to look at Dessert. I qdel'd Dessert, and after a moment, the camera suddenly jumped back to Ian.
> 
> And when tracking Ian, it wasn't even smooth. Just an instant jump. I saw the step changes to try and help the tracking, but it doesn't seem to work?

just added /event/destroyed to the thing that stops tracking

**Review by jwhitak — CHANGES_REQUESTED** [review]
Tested it again, this time simply tracked Salem. Camera followed normally. I moved away, getting the no longer tracking Salem message. Camera kept getting forced back to Salem every time the cat moved. So I tried tracking Ian. Salem was no longer getting tracked! But now the problem was centered on Ian.

**Comment by SECBATON-GRIFFON** [comment]
> Tested it again, this time simply tracked Salem. Camera followed normally. I moved away, getting the no longer tracking Salem message. Camera kept getting forced back to Salem every time the cat moved. So I tried tracking Ian. Salem was no longer getting tracked! But now the problem was centered on Ian.

should be fixed now

**Review by jwhitak — APPROVED** [review]

**Comment by SonixApache** [comment]
holy shit, at long last

**Comment by hacker-on-steroids** [comment]
toggle-camera-acceleration verb does nothing
not sure if it's caused by this or something else

**Pull request merged**